core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -19,9 +24,51 @@
+ * the name implies, all of the methods in this class must <b>never</b> block the invoking thread.
  * <p>
- * Implementations of this store must be thread safe if concurrent operations are performed on it. This should include
- * possibly invoking start or stop multiple times
+ * The first method that will be invoked on this store will be the {@link #start(InitializationContext)} to allow
+ * it to initialize and startup. Once the returned stage has completed the store is assumed to be in working condition
+ * for other operations to be performed upon it. Infinispan will guarantee visibility of variables written during
+ * the start method, so there is no need to synchronize these manually, unless they are mutated in the normal operations
+ * of the store itself.
+ * <p>
+ * Infinispan may at any point after this store has started query what characteristics this store has via
+ * {@link #characteristics()}. It is highly recommended that this method never change what values it returns once the
+ * store has been started as these may or may not be cached. For more information on how the characteristics affect
+ * the store operations, please see {@link Characteristic} and its various values.
+ * <p>
+ * By default this store only requires half a dozen or so methods to be implemented. However, there are more
+ * optional methods that may be implemented. If you implement such a method, please be sure to advertise the appropriate
+ * characteristic for that method, so Infinispan knows to invoke it. Note that if Infinispan has been told a
+ * characteristic is available and the method is not overridden, an {@link UnsupportedOperationException} will be
+ * thrown when trying to invoke the appropriate method. Each {@link Characteristic} defines what methods map to which
+ * characteristic.
+ * <p>
+ * All methods that are able include segment arguments in them, such as {@link #load(int, Object)} and
+ * {@link #publishEntries(IntSet, Predicate, boolean)}. It is not required, but recommended if possible, for a store
+ * to support segmentation. An Infinispan Cache can perform much more efficiently when segmentation is supported when
+ * performing bulk operations such as {@code Cache.size()} or {@code Cache.entrySet().stream()}. It also can very greatly
+ * increase state transfer time, when {@link PersistenceConfiguration#fetchPersistentState()} is enabled and also when
+ * removing data by segments. Ensure that it the store implementation supports segmentation that it returns the
+ * {@link Characteristic#SEGMENTABLE} characteristic when {@link #characteristics()} is invoked. A store implementation
+ * can tell if segmentation is enabled by checking the store configuration {@link StoreConfiguration#segmented()} available
+ * from the {@code InitializationContext}.
+ * <p>
+ * Some stores may have to interact with blocking APIs to perform their desired operation. Since we can never block
```

Comment:
"A store implementation may have to interact with blocking APIs to perform their required operations, however we should never block the invoking thread, therefore Infinispan provides a utility helper for these operations."